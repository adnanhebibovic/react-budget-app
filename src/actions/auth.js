import {firebase, provider} from '../services/firebase'

export const login = function(uid) {
    return {
        type: 'LOGIN',
        uid: uid
    }
}

export const startLogin = function() {
    return function() {
        return firebase.auth().signInWithPopup(provider)
    }
}

export const logout = function() {
    return {
        type: 'LOGOUT'
    }
}

export const startLogout = function() {
    return function() {
        return firebase.auth().signOut()
    }
}