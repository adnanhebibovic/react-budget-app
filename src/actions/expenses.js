import db from '../db/firebase'
import {v4 as uuid} from 'uuid'

export const addExpenseToStore = function(expense = {}) {
    const defaults = {
        id: new uuid(),
        title: '',
        amount: 0,
        createdAt: new Date().getTime(),
        ...expense
    };

    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: defaults.id,
            title: defaults.title,
            amount: defaults.amount,
            createdAt: defaults.createdAt
        }
    };
}

export const addExpense = function(expense = {}) {
    const defaults = {
        createdAt: new Date().getTime(),
        ...expense
    }

    return function(dispatch) {
        return db.collection('expenses').add(defaults)
        .then((doc) => {
            dispatch(addExpenseToStore({
                id: doc.id,
                ...defaults
            }))
        })
    }
}

export const removeExpenseFromStore = function(id) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

export const removeExpense = function(id) {
    return function(dispatch) {
        return db.collection('expenses').doc(id).delete()
        .then(() => {
            dispatch(removeExpenseFromStore(id))
        })
    }
}

export const editExpenseInStore = function(id, expense) {
    const defaults = {
        title: '',
        amount: 0,
        ...expense
    };

    return {
        type: 'EDIT_EXPENSE',
        id,
        expense: defaults
    }
}

export const editExpense = function(id, expense) {
    return function(dispatch) {
        return db.collection('expenses').doc(id).update(expense)
        .then(() => {
            dispatch(editExpenseInStore(id, expense))
        })
    }
}