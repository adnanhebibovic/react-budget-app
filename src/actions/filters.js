export const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

export const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

export const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

export const setDateFilters = function(startDate = undefined, endDate = undefined) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}
