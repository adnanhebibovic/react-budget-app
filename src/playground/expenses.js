import { createStore, combineReducers } from 'redux'
import uuid from 'uuid'

const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date()
        }
    }
}

const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}

const expensesDefaultState = []

const expensesReducer = function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}

const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

const setDateFilters = function({startDate = undefined, endDate = undefined} = {}) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}

const filtersDefaultState = {
    text: undefined,
    sortBy: undefined,
    startDate: undefined,
    endDate: undefined
}

const filtersReducer = function(state = filtersDefaultState, action) {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            return {
                ...state,
                sortBy: action.sortBy
            }
        case 'SET_DATE_FILTERS':
            return {
                ...state,
                ...action.dates
            }
        default:
            return state;
    }
}

const getExpenses = function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text === 'undefined' || expense.title.includes(text);
        const startDateMatch = typeof startDate === 'undefined' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate === 'undefined' || expense.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy)
    {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result
}

const store = createStore(
    combineReducers({
        expenses: expensesReducer,
        filters: filtersReducer
    })
)

store.dispatch(addExpense({
    title: 'January bills',
    amount: 279.00 
}))

store.dispatch(addExpense({
    title: 'February bills',
    amount: 523.00 
}))

store.dispatch(addExpense({
    title: 'March bills',
    amount: 358.00 
}))

store.dispatch(addExpense({
    title: 'New car',
    amount: 1000.00 
}))

store.subscribe(() => {
    const state = store.getState();
    console.log(getExpenses(state.expenses, state.filters))
})

store.dispatch(setTextFilter('bills'))
store.dispatch(sortByAmount())
store.dispatch(sortByDate())
store.dispatch(setDateFilters({endDate: new Date()}))
