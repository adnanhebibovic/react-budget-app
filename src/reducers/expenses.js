export const expensesDefaultState = []

export default function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'SET_EXPENSE':
            return state.concat(action.expenses.filter((expense) => state.findIndex((item) => item.id === expense.id) < 0))
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id === action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}