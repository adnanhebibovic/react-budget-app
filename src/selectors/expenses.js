export default function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text == 'undefined' || expense.title.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate == 'undefined' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate == 'undefined' || expense.createdAt < endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy) {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result;
}
