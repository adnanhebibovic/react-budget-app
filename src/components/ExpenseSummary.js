import React from 'react'
import {connect} from 'react-redux'
import numeral from 'numeral'
import { NavLink } from 'react-router-dom'

import {count, total} from '../selectors/expenses-statistics'
import select from '../selectors/expenses'

function ExpenseSummary(props) {
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{count(props.expenses)}</span> {count(props.expenses) === 1 ? 'expense' : 'expenses'} totalling <span>{numeral(total(props.expenses)).format('$0,0[.]00')}</span></h1>
                <div className="page-header__actions">
                    <NavLink className="button" to="/expenses">Add expense</NavLink>
                </div>
            </div>            
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: select(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)