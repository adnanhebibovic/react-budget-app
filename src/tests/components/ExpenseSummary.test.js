import React from 'react';
import { Provider } from 'react-redux';
import { create, act } from 'react-test-renderer';
import configureStore from '../../store/expense-store'

import expenses from '../fixtures/expenses';
import {removeExpenseFromStore} from '../../actions/expenses'
import ExpenseSummary from '../../components/ExpenseSummary';

let store;
let renderer;

beforeEach(() => {
    store = configureStore(expenses)
  
    renderer = create(
      <Provider store={store}>
        <ExpenseSummary />
      </Provider>
    );
});

test('Should render ExpenseSummary component correctly', () => {
    expect(renderer).toMatchSnapshot();
})

test('Should render ExpenseSummary component correctly with missing data', () => {
    act(() => {
        store.dispatch(removeExpenseFromStore(expenses[0].id));
    });
    
    act(() => {
        store.dispatch(removeExpenseFromStore(expenses[1].id));
    });

    act(() => {
        store.dispatch(removeExpenseFromStore(expenses[2].id));
    });

    act(() => {
        store.dispatch(removeExpenseFromStore(expenses[3].id));
    });

    expect(renderer).toMatchSnapshot();
})

