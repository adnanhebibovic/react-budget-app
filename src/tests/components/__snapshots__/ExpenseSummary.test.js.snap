// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseSummary component correctly 1`] = `
<div>
  <h2>
    Viewing 
    4
     
    expenses
     totalling 
    1,200.0
  </h2>
</div>
`;

exports[`Should render ExpenseSummary component correctly with missing data 1`] = `
<div>
  <h2>
    Viewing 
    0
     
    expenses
     totalling 
    0.0
  </h2>
</div>
`;
